src/libraries/Microsoft.Extensions.Configuration.Binder/tests/ConfigurationBinderTests.cs

```diff
@@ -112,6 +112,178 @@ public class DerivedOptionsWithIConfigurationSection : DerivedOptions
             public IConfigurationSection DerivedSection { get; set; }
         }
 
+        public record RecordTypeOptions(string Color, int Length);
```

Comment:
> This brings up the larger issue that we should probably not just be passing null into parameters that are missing unless they have a default value or are nullable.

I didn't see this earlier comment addressed. If I were to update the above line to the following, `CanBindRecordOptions` would still pass. `string Foo` would just be null despite having a non-nullable annotation.

```suggestion
#nullable enable
        public record RecordTypeOptions(string Color, int Length, string Foo);
#nullable restore
```

When we set properties, we can just skip over the property if config doesn't provide a value, but calling a constructor that we can observe requires a non-nullable parameter with a null argument seems very wrong.

Reply:
Yes, I can see the issue here. Part of me thinks "Let's just check if the property is nullable and throw if we don't have config for it", and another part thinks "The best effort approach is useful, but too dangerous, so get rid of it".

I'll got with a consensus here - is it worth getting another opinion? If not, I'll happily go with whatever you think.

Reply:
I think we will get feedback if we don't allow for default values. We respect them in DI.

Reply:
@eerhardt - sorry, are you saying that it's OK to allow defaults (nulls)?

Reply:
> sorry, are you saying that it's OK to allow defaults (nulls)?

I discussed offline with @eerhardt and we think we should match DI's behavior for default values, so the answer to your question is no unless the default value is null (e.g. `string Foo = null`). In any other case parameters are *required* even if the parameter is marked as non-nullable.

```csharp
public record RecordTypeOptions(string? Color, int Length, string Foo = "f");
```

In the above example, if the config is just `Length = 5` with no `Color = "blue"`, the binder should throw an `InvalidOperationException` even though `Color` has a nullable annotation. This would be similar to how we're throwing for `in/out/ref` constructor parameters, but the error message should make it very clear that the parameter is required but there's no matching config.


We can consider support using nullable annotations for optional constructor parameters later in a separate issue.

Reply:
I've implemented this. The first check sees if the parameters in the construct have a matching property. The next check, checks that each parameter has a matching config item.
I think the first check might be overkill though... What do you think?